### PR TITLE
YL GDPR Consent Framework implementation

### PR DESCRIPTION
**Pull request by anupash**
GDPR Consent Framework (https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework)
introduces vendor consent string. The consent of a user is expressed and
stored in consent string. The format is standardized and provided by
IAB.

Provides:
1. Renamed ConsentStringParser class to a more general GdprConsent class
2. Provided Builder method to create a consent String
3. Time-based Gdpr Switch

### CODE REVIEW

File: src/main/java/com/yieldlab/gdpr/GdprConsent.java [hunk 1]
```diff
@@ -0,0 +1,806 @@
...
+/**
+ * Copied from https://github.com/InteractiveAdvertisingBureau/Consent-String-SDK-Java and modified
+ *
+ * This class implements a parser for the IAB consent string as specified in
+ * https://github.com/InteractiveAdvertisingBureau/GDPR-Transparency-and-Consent-Framework/blob/master/
+ * Draft_for_Public_Comment_Transparency%20%26%20Consent%20Framework%20-%20cookie%20and%20vendor%20list%20format%
+ * 20specification%20v1.0a.pdf
+ *
+ */
+public class GdprConsent {
```
Comment (brushmate) [inline src/main/java/com/yieldlab/gdpr/GdprConsent.java]: Why did you rename the class?

File: gradlew.bat [hunk 2]
```diff
@@ -1,84 +1,84 @@
...
-rem Set variable GRADLE_EXIT_CONSOLE if you need the _script_ return code instead of
-rem the _cmd.exe /c_ return code!
-if  not "" == "%GRADLE_EXIT_CONSOLE%" exit 1
-exit /b 1
-
-:mainEnd
-if "%OS%"=="Windows_NT" endlocal
-
-:omega
+@if "%DEBUG%" == "" @echo off
```
Comment (twz123) [inline gradlew.bat:1]: Please leave CRLF line endings, since it's a windows file.

Might be worth adding a `.gitattributes` file like this:

```
* text=auto
*.bat !text
*.cmd !text
.gitignore !text
```

File: src/main/java/com/yieldlab/gdpr/GdprConsent.java (hunk 1 above)
Comment (twz123) [reply src/main/java/com/yieldlab/gdpr/GdprConsent.java]: I think it makes sense to separate the parsing logic into its own class and leave this one as a pure immutable value class?

File: src/main/java/com/yieldlab/gdpr/util/GdprSwitch.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+import static java.time.format.DateTimeFormatter.ISO_OFFSET_DATE_TIME;
+
+import java.time.Instant;
+import java.util.Optional;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.yieldlab.gdpr.GdprConsent;
+
+public class GdprSwitch {
```
Comment (twz123) [inline src/main/java/com/yieldlab/gdpr/util/GdprSwitch.java]: Not sure if all the methods in this class are really helpful. The only thing in which we're interested is the switch time. `isVendorAllowed` seems to be very special, after all.

Comment (nabcos) [reply src/main/java/com/yieldlab/gdpr/util/GdprSwitch.java]: And if we want to contribute back, `isOn` should not be here but in the app IMO, as this library should only be about the consent string.

File: build.gradle [hunk 4]
```diff
@@ -1,12 +1,49 @@
...
+    }
+}
+
+uploadArchives {
+    repositories {
+        mavenDeployer {
+            repository(url: "file://~/.m2/")
+            pom.version = '0.0.1-SNAPSHOT'
+            pom.artifactId = 'gdpr-consent-framework'
+            pom.groupId = 'com.yieldlab'
```
Comment (nabcos) [inline build.gradle]: If we want to contribute back, this should not be `com.yieldlab`, but I guess for the IAB this is not set up at all - so we'll fix that later if needed.

File: src/main/java/com/yieldlab/gdpr/GdprConsent.java [hunk 5]
```diff
@@ -0,0 +1,806 @@
...
+                } else {
+                    int vendorId = bits.getInt(currentOffset, VENDOR_ID_SIZE);
+                    currentOffset += VENDOR_ID_SIZE;
+                    rangeEntries.add(new RangeEntry(vendorId));
+                }
+            }
+        }
+    }
+
+    public GdprConsent(Builder builder) throws VendorConsentException {
```
Comment (nabcos) [inline src/main/java/com/yieldlab/gdpr/GdprConsent.java]: Maybe private and only accessible via `Builder#build`?

**Comment by nabcos** [comment]
There was a general comment from ActiveAgent guys that consent parsing was rather slow, adding more than few ms to their processing time. @anupash did you measure that already?

**Comment by nabcos** [comment]
Just FYI there's another PR to add building consent string:
https://github.com/InteractiveAdvertisingBureau/Consent-String-SDK-Java/pull/18

File: src/main/java/com/yieldlab/gdpr/GdprConsent.java (hunk 1 above)
Comment (anupash) [reply src/main/java/com/yieldlab/gdpr/GdprConsent.java]: I have restored the name of the class on popular request.

File: gradlew.bat (hunk 2 above)
Comment (anupash) [reply gradlew.bat:1]: done

File: src/main/java/com/yieldlab/gdpr/util/GdprSwitch.java (hunk 3 above)
Comment (anupash) [reply src/main/java/com/yieldlab/gdpr/util/GdprSwitch.java]: I have removed this file from the project and inlined it in adserver.

File: build.gradle (hunk 4 above)
Comment (anupash) [reply build.gradle]: i can set it back to com.iab if we wish to contribute to IAB

**Comment by anupash** [comment]
@nabcos I have not performed any performance tests yet. But plan to do so maybe later.

**Comment by anupash** [comment]
@nabcos i raised a PR to the IAB consent string sdk and it was accepted 😄, albeit a little too quickly.
